### DISCUSSION TITLE
Individual Low Cell Masking for the Visualize Tab

### DISCUSSION THREAD

**Discussion by Ram-SN** [uwrit/leaf#571] [Ideas]
We have a Use Case at CU Anschutz for compliance to block the visualize tab when individual values are 10 or less. 
<img width="1600" alt="Screen Shot 2022-08-25 at 6 04 21 PM" src="https://user-images.githubusercontent.com/56275846/191068151-927eab77-08d3-4239-823b-72982530aa7e.png">

We implemented a fix to visualize.tsx:

```
if (
            cohort.visualization.demographics.ageByGenderData.buckets['<1'].females <= 9 &&
            cohort.visualization.demographics.ageByGenderData.buckets['<1'].females > 0 ||
            cohort.visualization.demographics.ageByGenderData.buckets['1-9'].females <= 9 &&
            cohort.visualization.demographics.ageByGenderData.buckets['1-9'].females > 0 ||
            cohort.visualization.demographics.ageByGenderData.buckets['18-24'].females <= 9 &&
            cohort.visualization.demographics.ageByGenderData.buckets['18-24'].females > 0 ||
            cohort.visualization.demographics.ageByGenderData.buckets['25-34'].females <= 9 &&
            cohort.visualization.demographics.ageByGenderData.buckets['25-34'].females > 0 ||
            cohort.visualization.demographics.ageByGenderData.buckets['35-44'].females <= 9 &&
            cohort.visualization.demographics.ageByGenderData.buckets['35-44'].females > 0 ||
            cohort.visualization.demographics.ageByGenderData.buckets['45-54'].females <= 9 &&
            cohort.visualization.demographics.ageByGenderData.buckets['45-54'].females > 0 ||
            cohort.visualization.demographics.ageByGenderData.buckets['55-64'].females <= 9 &&
            cohort.visualization.demographics.ageByGenderData.buckets['55-64'].females > 0 ||
            cohort.visualization.demographics.ageByGenderData.buckets['65-74'].females <= 9 &&
            cohort.visualization.demographics.ageByGenderData.buckets['65-74'].females > 0 ||
            cohort.visualization.demographics.ageByGenderData.buckets['75-84'].females <= 9 &&
            cohort.visualization.demographics.ageByGenderData.buckets['75-84'].females > 0 ||
            cohort.visualization.demographics.ageByGenderData.buckets['>84'].females <= 9 &&
            cohort.visualization.demographics.ageByGenderData.buckets['>84'].females > 0
        ){
            return <CellTooSmallBox cacheLimit={10}/>
        }
```

We implemented similar hacky fixes for all components and we are working on a full patch to remove this blocky code from the React App. This is what we have discovered so far:

![image](https://user-images.githubusercontent.com/56275846/191074630-0ae99fd1-eea2-440c-9a64-2b236385bb35.png)

**Comment by ndobb** [comment]
What is your proposal, @Ram-SN ? A user can do simple arithmetic and count that 315 - (other values that sum to 315 besides number of interest) = a number less than 10. So this is not an easy problem to solve.

**Reply by Ram-SN** [reply]
We're working on a complete message, will update once done

**Reply by Ram-SN** [reply]
We updated the discussion, any feedback is welcome thanks!

**Reply by glabrie10** [reply]
@ndobb this patch will block the visualize tab if any individual cell is less than 10. No masking. This blocks the user from deriving values.

**Comment by ndobb** [comment]
I'm really not sure on this one @Ram-SN @glabrie10 . What **precisely** is the desired behavior here? To simply not show a number less than 10, even when users can reason it out without too much difficulty?

@glabrie10  per your last comment, I understand maybe you are moving instead to simply disable the Visualize tab under certain circumstances. Is that correct? I'm thinking that (1) doing so may lead to some major user confusion, as the Visualize tab may appear to randomly be visible some times and not others, and (2) ultimately if the API is still sending the underlying data, any somewhat-technical user could see the real numbers by just pulling up the dev console, so unless the API is changed it's not really a robust solution.

**Comment by glabrie10** [comment]
@ndobb We would like to disable visualization whenever any value used to render graphics is ten or lower. We have created a new custom error message that explains in detail why visualization is disabled for cohorts with any value of 10 or lower. The API is part of the entire patch. We have not worked on that. @Ram-SN, our engineer on this patch, is still working through discovery. We understand that the API will need patching before this is considered a viable solution. @ndobb, if you have any insight on what files to look at for the API, we would greatly appreciate it.
